File: python/more/search-order-download.py
#!/usr/bin/env python3

# In this example we search for a company, place an order for a report, wait for the report to be ready and then download it

import requests
import json
import time

base_rest_url = "https://test.cofacecentraleurope.com/api/bi/v1/"

headers = {'accept': 'application/json', 'content-type': 'application/json', 'charset': 'UTF-8', 'api_key': '8842ace2-e377-48d9-b129-f952950ea535'}


def search_for_company(params):
    print("")
    print("searching for company...")
    response = requests.get(base_rest_url + "companies?" + params, headers=headers)
    if (response.status_code == 200):
        return response.json()
    else:
        raise ValueError("server returned http status %d (%s)" % (response.status_code, response.text))


def get_order_details(id):
    print("")
    print("fetching order details...")
    response = requests.get(base_rest_url + "order/" + id, headers=headers)
    if (response.status_code == 200):
        return response.json()
    else:
        raise ValueError("server returned http status %d (%s)" % (response.status_code, response.text))


def list_order_documents(id):
    print("")
    print("fetching list of order documents...")
    response = requests.get(base_rest_url + "order/" + id + "/documents", headers=headers)
    if (response.status_code == 200):
        return response.json()
    else:
        raise ValueError("server returned http status %d (%s)" % (response.status_code, response.text))


def get_order_document(id, file_name):
    print("")
    print("fetching order document...")
    response = requests.get(base_rest_url + "order/" + id + "/document/" + file_name, headers=headers)
    if (response.status_code == 200):
        return response.content
    else:
        raise ValueError("server returned http status %d (%s)" % (response.status_code, response.text))


def place_order_immediate(product, company_id, format):
    print("")
    print("placing immediate order...")
    data = json.dumps(
        {
            "customer_reference": "testref-313",
            "company_id": company_id,
            "product_details": {
                "product_code": product,
                'language': 'EN',
                'format': format,
            },
            "credit_report_details": {
                'research_instructions': 'immediate_no_research'
            },
        })
    response = requests.post(base_rest_url + "orders/creditreport", headers=headers, data=data)
    if (response.status_code == 202):
        return response.json()
    else:
        raise ValueError("server returned http status %d (%s)" % (response.status_code, response.text))


def print_title(title):
    print("\n\n===== " + title + " =====\n")


print_title("Company search")

companies = search_for_company("country_iso_code=PL&company_name=Hydrobudowa&limit=5")
print("Company search result list:")
for company in companies:
    print("  {0}: {1}".format(company['company_id'], company['company_name']))
company_id = companies[0]['company_id']
print("")
print("Found {0} companies with name Hydrobudowa in PL; first one has company_id {1}".format(len(companies), company_id))

print_title("Placing an order")

immediate_order_response = place_order_immediate(200, company_id, 'html')
immediate_order_guid = immediate_order_response['id']
print("")
print("Immediate order placed; order ID is " + immediate_order_guid)


print_title("Waiting for the order to finish")

finished = False
while not finished:
    print("Order not finished yet, waiting...")
    time.sleep(5)
    order_details = get_order_details(immediate_order_guid)
    finished = (order_details['order_status'] == 'finished' or order_details['order_status'] == 'cancelled')

print("Immediate order finished")

print_title("Retrieving report document")

documents = list_order_documents(immediate_order_guid)
print("")
print("Order documents list:")
for document in documents:
    print("  {0}".format(document['file_name']))

report = get_order_document(immediate_order_guid, documents[0]['file_name'])
print("")
print("Downloaded report has a length of {0} characters".format(len(report)))
print(report)
print("...")

File: report-api/v1/sample/C#/complete-workflows/search-order-download.cs
using System;
using System.Net.Http;
using Newtonsoft.Json.Linq;

// In this example we search for a company, place an order for a report, wait for the report to be ready and then download it

namespace ConsoleProgram
{
    public class Class1
    {
        private const string base_rest_url = "https://test.cofacecentraleurope.com/api/bi/v1/";

        private static JArray SearchForCompany(HttpClient httpClient, string queryParams)
        {
            var response = httpClient.GetStringAsync(new Uri(base_rest_url + "companies?" + queryParams)).Result;
            return JArray.Parse(response);
        }

        private static JObject PlaceOrderImmediate(HttpClient httpClient, string product_code, string company_id, string format)
        {
            string payload = "{\"company_id\": \"" + company_id + "\",\"product_details\": {\"product_code\": \"" + product_code + "\",\"language\": \"EN\",\"format\": \"" + format + "\"},\"credit_report_details\": {\"research_instructions\": \"immediate_no_research\"}}";
            HttpContent content = new StringContent(payload, System.Text.Encoding.UTF8, "application/json");
            var response = httpClient.PostAsync(new Uri(base_rest_url + "orders/creditreport"), content).Result;
            return JObject.Parse(response.Content.ReadAsStringAsync().Result);
        }

        private static JObject GetOrderDetails(HttpClient httpClient, string order_guid)
        {
            var response = httpClient.GetStringAsync(new Uri(base_rest_url + "order/" + order_guid)).Result;
            return JObject.Parse(response);
        }

        private static JArray ListOrderDocuments(HttpClient httpClient, string order_guid)
        {
            var response = httpClient.GetStringAsync(new Uri(base_rest_url + "order/" + order_guid + "/documents")).Result;
            return JArray.Parse(response);
        }

        private static string GetOrderDocument(HttpClient httpClient, string order_guid, string file_name)
        {
            var response = httpClient.GetStringAsync(new Uri(base_rest_url + "order/" + order_guid + "/document/" + file_name)).Result;
            return response.ToString();
        }

        static void Main(string[] args)
        {
            using (var httpClient = new HttpClient())
            {
                httpClient.DefaultRequestHeaders.Add("api_key", "8842ace2-e377-48d9-b129-f952950ea535");
                httpClient.DefaultRequestHeaders.Add("accept", "application/json");

                JArray companies = SearchForCompany(httpClient, "country_iso_code=PL&company_name=Hydrobudowa&limit=5");

                Console.WriteLine("{0} companies found", companies.Count);

                string company_id = companies[0]["company_id"].ToString();
                Console.WriteLine("Found {0} companies with name Hydrobudowa in PL; first one has company_id {1}", companies.Count, company_id);

                Console.WriteLine("Placing an order");

                JObject immediate_order_response = PlaceOrderImmediate(httpClient, "200", company_id, "html");
                var immediate_order_guid = immediate_order_response["id"].ToString();
                Console.WriteLine("Immediate order placed; order ID is " + immediate_order_guid);

                var finished = false;
                while (!finished)
                {
                    Console.WriteLine("Order not finished yet, waiting...");
                    System.Threading.Thread.Sleep(5000);
                    JObject order_details = GetOrderDetails(httpClient, immediate_order_guid);
                    string order_status = order_details["order_status"].ToString();
                    finished = order_status == "finished" || order_status == "cancelled";
                }

                Console.WriteLine("Immediate order finished, listing documents");

                JArray documents = ListOrderDocuments(httpClient, immediate_order_guid);

                foreach (var document in documents.Children())
                {
                    string file_name = document["file_name"].ToString();
                    Console.WriteLine(file_name);
                    string report = GetOrderDocument(httpClient, immediate_order_guid, file_name);
                    Console.WriteLine(report);
                }
            }
        }
    }
}
